doc/source/serve/production-guide/fault-tolerance.md

```diff
@@ -0,0 +1,304 @@
+(serve-e2e-ft)=
+# Adding End-to-End Fault Tolerance
+
+Serve handles component failures _within_ a Ray cluster by default. You can configure your Ray cluster and Serve application to guard them against cluster-level failures as well.
+
+This section helps you:
+
+* provide additional fault tolerance for your Serve application
+* understand Serve's recovery procedures
+* simulate system errors in your Serve application
+
+:::{admonition} Relevant Guides
+:class: seealso
+This section discusses concepts from:
+* Serve's [architecture guide](serve-architecture)
+* Serve's [Kubernetes production guide](serve-in-production-kubernetes)
+:::
+
+(serve-e2e-ft-guide)=
+## Guide: end-to-end fault tolerance for your Serve app
+
+Serve provides some [fault tolerance](serve-ft-detail) features out of the box. You can provide end-to-end fault tolerance by tuning these features and running Serve on top of [KubeRay].
+
+### Replica health-checking
+
+By default, the Serve controller periodically health-checks each Serve deployment replica and restarts it on failure.
+
+You can define custom application-level health-checks and adjust their frequency and timeout.
+To define a custom health-check, add a `check_health` method to your deployment class.
+This method should take no arguments and return no result, and it should raise an exception if the replica should be considered unhealthy.
+The Serve controller logs the raised exception if the health-check fails.
+You can also use the deployment options to customize how frequently the health-check is run and the timeout after which a replica is marked unhealthy.
+
+```{literalinclude} doc_code/fault_tolerance/replica_health_check.py
+:start-after: __health_check_start__
+:end-before: __health_check_end__
+:language: python
+```
+
+### Worker node recovery
+
+:::{admonition} KubeRay Required
+:class: caution, dropdown
+You **must** deploy your Serve application with [KubeRay] to use this feature.
+
+See Serve's [Kubernetes production guide](serve-in-production-kubernetes) to learn how you can deploy your app with KubeRay.
+:::
+
+By default, Serve can recover from certain failures _within_ a Ray cluster, such as unhealthy actors. When [Serve runs on Kubernetes](serve-in-production-kubernetes) with [KubeRay], it can also recover from some _cluster-level_ failures, such as dead worker nodes.
```

Comment:
If this is standard distributed systems terminology, feel free to ignore--but it seems a little confusing to me that a node failure is not considered a failure within a Ray cluster.  Maybe we need some other way to refer to failures "within a Ray cluster".  Is "Ray-level failures" accurate?

Reply:
No, it's not standard terminology. I borrowed it from [this doc page](https://docs.ray.io/en/releases-2.0.0/serve/production-guide/failures.html#failure-recovery). I don't think the distinction is too critical for the user's understanding, so I removed that wording.